contributors/design-proposals/storage/raw-block-ro.md

```diff
@@ -0,0 +1,139 @@
+# Raw Block Read-Only Consumption in Kubernetes
+
+Authors: erinboyd@, screeley44@, mtanino@
+
+This document presents a proposal for honoring the read-only setting for raw block devices in the CRI into Kubernetes. This document builds on the design
+outlined in #805 & #1265
+
+# Terminology
+* Raw Block Device - a physically attached device devoid of a filesystem
+* Raw Block Volume - a logical abstraction of the raw block device as defined by a path
+* Filesystem on Block - a formatted (ie xfs) filesystem on top of a raw block device
+
+# Goals
+* Create consistent interaction with container runtime for readOnly block devices 
+
+  
+# Specific use case of storage consumption for read-only:
+  * An admin wishes to expose a block volume for read-only consumption to an application without write permissions. Example
+ applications might be a database with data to be used for analytics processing in which the results of processing are written to another fileystem or application.  
+
+ 
+# Design Overview
+
+ Today in the container runtime, the raw block devices are defaulted to 'rwm' which provides, "read", "write" and "mknod". There exists such use cases as
+outlined above in which we want to restrict the user to only access the block device as a reader. This support already exists in the container runtime and should
+be supported in Kube. 
+In addition, in the container spec today we use the 'readOnly' parameter for the same purpose when using a filesystem. The proposal is to add this same parameter to the volumeDevices definition in the API.
+The PVs generated from setting this device flag should be set to RO in the accessModes to facilitate proper binding. Though this is not enforced in Kube today is should be seen as a best practice so the user is aware of different modes of storage consumption.
+ 
+# Proposed API Changes
+   
+## Pod API Changes:
+This change adds the 'readOnly' flag, conistent with the filesystem implementation in volumeMounts, to the volumeDevices definition.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-db
+spec:
+    containers:
+    - name: mysql
+      image: mysql
+      volumeDevices: 
+      - name: my-db-data
+        devicePath: /dev/xvda 
+        readOnly: true #proposed API change
+    volumes:
+    - name: my-db-data
+      persistentVolumeClaim:
+	claimName: raw-pvc
+```
+
+# Use Case
+
+## UC1: 
+
+DESCRIPTION: An admin wishes to restrict usage of a raw block device to read-only. 
+
+WORKFLOW:
+Admin creates PV mapping to raw block device set to read-only permissions.
+
+User creates pod to consume block device and specifies readOnly permissions on the device.
+
+ADMIN:
+
+```
+kind: PersistentVolume
+apiVersion: v1
+metadata:
+  name: local-raw-pv
+spec:
+  volumeMode: Block
+  capacity:
+    storage: 100Gi
+  local:
```

Comment:
for the Admin part in this, they only control the PV in this case and therefore we should either comment that ```If readOnly property is available for the VolumeSource it can be set to true to make the device readOnly``` Or choose a volumeSource that already has the readOnly property.  Does that make sense?

Reply:
We already have issues with PV volume sources that have a readonly field which is ignored in favor of the PV spec readonly field. This is adding a third level. What's the precedence?

Reply:
@liggitt this is following the same pattern already set forth for filesystems. When using a filesystem in the container/pod spec we have volumeMounts and here is where we specify the readOnly. We are just extending the same to block so they will look/feel the same to the user.

Reply:
the readOnly field set on the volumeMount is ignored in favor of the PVCVolumeSource readOnly field today.

Reply:
@liggitt @screeley44 

Can I confirm that current filesystem volume's situation of readOnly parameter?
This is my current understanding of readOnly parameter.

object | effect | who set this?
------------ | ------------- | ------------- 
readOnly in VolumeSource | this parameter is valid for inline volume. If volume is PV, this parameter is ignored. | Admin
readOnly in PesistentVolumeSource | this parameter is ignored even if volume is PV | Admin
readOnly in volumeMounts | In the volume plugins, they check the field and then potentially chmod all the directories/files in the volume based on that | User
readOnly in PersistentVolumeClaimVolumeSource | this parameter is valid for PV | user

When we add readOnly parameter handling for block volume, I think we should add same behaviour to avoid admin/users confusion.

Reply:
@liggitt there is no RO field in either the PV or the PVC, only the accessMode. the RO field today only exists in the pod spec

Reply:
Ok, there's 2 places today:
* plugin specific volume source (including PVC source)
* volumeMount

The readOnly flag in the plugin source overrides the volumeMount

Reply:
ok, so VolumeDevice should behave the same then right?  Meaning we provide a readOnly flag in the VolumeDevice() same as VolumeMount() and if a plugin specific volume source (including PVC source) has the readOnly flag it should override the VolumeDevice  ... is this correct @msau42 @erinboyd @mtanino @liggitt ?